### DISCUSSION TITLE
When should you not use the `next/dynamic`?

### DISCUSSION THREAD

**Discussion by jfarris587** [vercel/next.js#32895] [Help]
`dynamic` importing helps with smaller initial load JS bundles which helps performance, which helps SEO, and other things. But when should you NOT use `next/dynamic`? 

**Here are some scenarios that '_seem bad'_ but I don't have enough background to justify _'why'_**

### Too many extra JS bundles?

When using `dynamic/import` over and over, you are adding extra bundles that the application will have to fetch. This is fine, if they aren't fetched until a user interaction, but what if they are instantly loaded after the first render?

For example, lets say all your `dynamic` imports are just normal components. You don't dynamically import 'on user click', instead the app will import them on first client render. So all those extra JS bundles are immediately fetched.

**could this be bad if there are too many? how many is 'too many' JS files to request?**

### Why not always dynamic import components with `{ ssr: true }`

The biggest reason to not dynamic import is because its needed on first load, especially for SSR. Say, big components above the fold, or content needed for SEO purposes. It has to be served rendered by the server first. 

But with `{ ssr: true }` option:

```tsx
const MyComponent = dynamic(() => import('./MyComponent'), { ssr: true })
```

We can render the component on the server side, then still `dynamic` import it. So its Rendered, but we then fetch the additional JS needed for functionality (what happens when a user clicks, hovers, interacts, etc).

So now, the only problem is that the component wouldn't be interactive for the amount of time it takes for it to render on the client the first time, until the additional bundle is fetched via `dynamic` import.... but this is generally milliseconds.

**So why not default to having most if not all components be `{ssr: true}` dynamic imports for SSR, have the initial bundle be small, then fetch all the bundles associated with the components?**

**Comment by joaogarin** [comment]
This is an interesting point I think could worth a bit more discussion. I am considering also this for the purposes of achieving some sort of feature flag internal system where depending if flags are enabled the component is imported or not. and thinking if using dynamic import for these components makes sense or not.

See the discussion https://github.com/vercel/next.js/discussions/35358

Would be good to know more about the drawbacks of this approach and if indeed there is "too much dynamic import with ssr true" or not

**Comment by icyJoseph** [comment]
I can think of a few reasons to avoid ALWAYS splitting otherwise monolithic asynchronous work, into, multi-part bundles of asynchronous work. Not only applies to `dynamic` in Next, but `Suspense`. React 18 offers interesting solutions to this, but some of the problems are just inherit to code splitting. In general, these tools ought to be used responsibly. Bundling code for an entire route that the visitor is not using into the current route bundle is, of course, undesirable, but Next.js' routing handles this for you.
 
- One get request for `/path` splits into several to fetch the necessary bundles
- When presented with many parallel requests, the browser might delay some, meaning the UI won't necessarily come together all at once
- Fetching the bundles is asynchronous, it might fail, or take very long
- You make assumptions about how Search Engines process your site, will they wait for your dynamic components?
- If the server resolves to an HTML string, the client's first render must be exactly the same, in order to be able to render the DOM and attach event listeners proper. In the face of asynchronous behaviour this turns into a very difficult task. This is why you can't tell the server to assume a component is present, but send it asynchronously to the client.
- Additional bundles have a small overhead, which means, that a bundle made up of `(A + B)`, when split, results in `A + ( B + overhead)`, so it is not necessarily less total JavaScript, it's just split to cut short A's download/parsing.
- Layout shifts accumulation is undesirable, and proper handling increases code complexity, (`loading` option on `dynamic`)

**Reply by joaogarin** [reply]
Right, this makes sense. I still have some issues in relation to https://github.com/vercel/next.js/discussions/35358 but that is a slightly different topic. What you mention here to me makes a lot of sense, thanks for the feedback :+1: =)

**Reply by DaniloMladenovic** [reply]
Could you give some example of when next/dynamic should be used?
